File: ObjectCDemo/DatingApp/DatingApp/Controller/EventDetailViewController/EventDetailViewController.m
//
//  EventDetailViewController.m
//  DatingApp
//
//  Created by Steven Jimy   on 11/11/16.
//  Copyright © 2016 Steven Jimy . All rights reserved.
//

#import "EventDetailViewController.h"
#import "EventSlideTableViewCell.h"
#import "EventInforTableViewCell.h"
#import "EventGuestTableViewCell.h"
#import "GuestFactory.h"

typedef enum : NSUInteger {
    kEventDetailSlideCellIndex = 0,
    kEventDetailInforCellIndex,
    kEventDetailGuestCellIndex,
    kEventDetailNumberOfRow,
} EventDetailCellIndex;

@interface EventDetailViewController ()
<UITableViewDelegate, UITableViewDataSource, GuestFactoryDelegate>
{
    __weak IBOutlet UITableView *eventDetailTableView;
    __weak IBOutlet UIButton *joinButton;
    __weak IBOutlet UIButton *maybeButton;
    __weak IBOutlet UIButton *cancelButton;
    NSMutableArray *cellArray;
    __weak IBOutlet UIActivityIndicatorView *indicatorView;
    NSMutableArray *guestArray;
}
@end

@implementation EventDetailViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    [indicatorView startAnimating];
    [self initCells];
    [self getData];
}

#pragma mark - Get Data
- (void)getData{
    GuestFactory *guestFactory = [[GuestFactory alloc] init];
    [guestFactory getGuestOfEvent:self.eventModel withDelegation:self];
}

#pragma mark - GuestFactoryDelegate
- (void)getGuestOfEventWithSuccessHandler:(NSArray *)results{
    guestArray = [[NSMutableArray alloc] initWithArray:results];
    dispatch_async(dispatch_get_main_queue(), ^{
        [eventDetailTableView reloadData];
        [indicatorView stopAnimating];
    });
}

- (void)getGuestOfEventWithFailureHandler:(NSError *)error{
    guestArray = [NSMutableArray new];
    dispatch_async(dispatch_get_main_queue(), ^{
        [eventDetailTableView reloadData];
        [indicatorView stopAnimating];
    });
}

- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

#pragma mark - UITableViewDataSource
- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView{
    return 1;
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section{
    return kEventDetailNumberOfRow;
}

- (CGFloat)tableView:(UITableView *)tableView heightForRowAtIndexPath:(NSIndexPath *)indexPath{
    switch (indexPath.row) {
        case kEventDetailSlideCellIndex:
            return 150;
            break;
        case kEventDetailInforCellIndex:
            return 150;
            break;
        case kEventDetailGuestCellIndex:
            return 90;
            break;
        default:
            break;
    }
    return 0;
}

- (void)initCells{
    cellArray = [NSMutableArray new];
    static NSString *cellSlideIdentifier = @"eventCellSlideIdentifier";
    static NSString *cellInforIdentifier = @"eventCellInforIdentifier";
    static NSString *cellGuestIdentifier = @"eventCellGuestIdentifier";
    for (NSInteger i = 0; i < kEventDetailNumberOfRow; i++) {
        switch (i) {
            case kEventDetailSlideCellIndex:{
                EventSlideTableViewCell *cell = (EventSlideTableViewCell *)[eventDetailTableView dequeueReusableCellWithIdentifier:cellSlideIdentifier];
                if (cell == nil){
                    NSArray *nib = [[NSBundle mainBundle] loadNibNamed:@"EventSlideTableViewCell" owner:self options:nil];
                    cell = [nib objectAtIndex:0];
                }
                [cellArray addObject:cell];
            }
                break;
            case kEventDetailInforCellIndex:{
                EventInforTableViewCell *cell = (EventInforTableViewCell *)[eventDetailTableView dequeueReusableCellWithIdentifier:cellInforIdentifier];
                if (cell == nil){
                    NSArray *nib = [[NSBundle mainBundle] loadNibNamed:@"EventInforTableViewCell" owner:self options:nil];
                    cell = [nib objectAtIndex:0];
                }
                [cellArray addObject:cell];
            }
                break;
            case kEventDetailGuestCellIndex:{
                EventGuestTableViewCell *cell = (EventGuestTableViewCell *)[eventDetailTableView dequeueReusableCellWithIdentifier:cellGuestIdentifier];
                if (cell == nil){
                    NSArray *nib = [[NSBundle mainBundle] loadNibNamed:@"EventGuestTableViewCell" owner:self options:nil];
                    cell = [nib objectAtIndex:0];
                }
                [cellArray addObject:cell];
            }
                break;
            default:
                break;
        }
    }
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath{
    EventSlideTableViewCell *cell = [cellArray objectAtIndex:indexPath.row];
    switch (indexPath.row) {
        case kEventDetailSlideCellIndex:{
            EventSlideTableViewCell *cell = [cellArray objectAtIndex:kEventDetailSlideCellIndex];
            return cell;
        }
            break;
        case kEventDetailInforCellIndex:{
            EventInforTableViewCell *cell = [cellArray objectAtIndex:kEventDetailInforCellIndex];
            return cell;
        }
            break;
        case kEventDetailGuestCellIndex:{
            EventGuestTableViewCell *cell = [cellArray objectAtIndex:kEventDetailGuestCellIndex];
            [cell reloadWithGuestArray:guestArray];
            return cell;
        }
            break;
        default:
            break;
    }
    return cell;
}

#pragma mark - Actions
- (IBAction)touchJoinButton:(id)sender {
}

- (IBAction)touchMaybeButton:(id)sender {
}

- (IBAction)touchCancelButton:(id)sender {
}

- (IBAction)touchBackButton:(id)sender {
    [self.navigationController popViewControllerAnimated:YES];
}
@end

File: SwiftDemos/DatingApp/DatingApp/Controller/EventDetailViewController/EventDetailViewController.swift
//
//  EventDetailViewController.swift
//  DatingApp
//
//  Created by Steven Jimy   on 11/10/16.
//  Copyright © 2016 Steven Jimy . All rights reserved.
//

import UIKit

class EventDetailViewController: BaseViewController, UITableViewDelegate, UITableViewDataSource, UserFactoryDelegate {
    @IBOutlet weak var eventDetailTableView: UITableView!
    let cellArray : NSMutableArray = []
    var guestArray : NSMutableArray = []
    var eventModel : EventModel!
    
    
    // MARK : Button action
    @IBAction func touchBackButton(_ sender: AnyObject) {
        if let navController = self.navigationController {
            navController.popViewController(animated: true)
        }
    }

    override func viewDidLoad() {
        super.viewDidLoad()
        initCells()
        getData()
    }
    
    func getData() {
        let userFactory : UserFactory = UserFactory()
        userFactory.getGuestsOfEvent(eventId: eventModel.eventId, delegation: self)
    }
    
    // MARK : UserFactoryDelegate
    func getGuestsErrorHandler(error: NSError) {
        guestArray = NSMutableArray()
        DispatchQueue.main.async {
            self.eventDetailTableView.reloadData()
        }
    }
    
    func getGuestsSuccessHandler(guests: NSArray) {
        guestArray = NSMutableArray(array: guests)
        DispatchQueue.main.async {
            self.eventDetailTableView.reloadData()
        }
    }

    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
        // Dispose of any resources that can be recreated.
    }
    
    // MARK : UITableViewDataSource
    func numberOfSections(in tableView: UITableView) -> Int {
        return 1
    }
    
    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return 3
    }
    
    func tableView(_ tableView: UITableView, heightForRowAt indexPath: IndexPath) -> CGFloat {
        switch indexPath.row {
        case 0:
            return 150
        case 1:
            return 150
        case 2:
            return 90
        default:
            return 0
        }
    }
    
    func initCells() {
        let cellEventSlideIdentifier : String = "cellEventSlideIdentifier"
        let cellEventInforIdentifier : String = "cellEventInforIdentifier"
        let cellEventGuestsIdentifier : String = "cellEventGuestsIdentifier"
        
        
        let cellSlideNib : UINib = UINib(nibName: "EventSlideTableViewCell", bundle: nil)
        eventDetailTableView.register(cellSlideNib, forCellReuseIdentifier: cellEventSlideIdentifier)
        let cellSlide : EventSlideTableViewCell = eventDetailTableView.dequeueReusableCell(withIdentifier: cellEventSlideIdentifier) as! EventSlideTableViewCell
        cellArray.add(cellSlide)
        
        let cellInforNib : UINib = UINib(nibName: "EventInforTableViewCell", bundle: nil)
        eventDetailTableView.register(cellInforNib, forCellReuseIdentifier: cellEventInforIdentifier)
        let cellInfor : EventInforTableViewCell = eventDetailTableView.dequeueReusableCell(withIdentifier: cellEventInforIdentifier) as! EventInforTableViewCell
        cellArray.add(cellInfor)
        
        let cellGuestNib : UINib = UINib(nibName: "EventGuestsTableViewCell", bundle: nil)
        eventDetailTableView.register(cellGuestNib, forCellReuseIdentifier: cellEventGuestsIdentifier)
        let cellGuest : EventGuestsTableViewCell = eventDetailTableView.dequeueReusableCell(withIdentifier: cellEventGuestsIdentifier) as! EventGuestsTableViewCell
        cellArray.add(cellGuest)
    }
    
    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        switch indexPath.row {
        case 0:
            let cell : EventSlideTableViewCell = cellArray.object(at: 0) as! EventSlideTableViewCell
            return cell
        case 1:
            let cell : EventInforTableViewCell = cellArray.object(at: 1) as! EventInforTableViewCell
            return cell
        case 2:
            let cell : EventGuestsTableViewCell = cellArray.object(at: 2) as! EventGuestsTableViewCell
            cell.reloadView(guests: guestArray)
            return cell
        default:
            let cell : EventSlideTableViewCell = cellArray.object(at: 0) as! EventSlideTableViewCell
            
            return cell
        }
    }
}
